packages/asset-swapper/src/network/sassy_sampler.ts

```diff
@@ -0,0 +1,295 @@
+        sources: ERC20BridgeSource[],
+        takerToken: Address,
+        makerToken: Address,
+        takerAmount: BigNumber,
+    ): Promise<DexSample<TwoHopFillData>[]> {
+        const tokenPaths = this._getTwoHopTokenPaths(takerToken, makerToken);
+        const hopResults = (await Promise.all(tokenPaths.map(async tokenPath =>
+            this._twoHopSampler.getTwoHopSellSampleAsync(sources, tokenPath, takerAmount))));
+        return hopResults.filter(h => !!h) as DexSample<TwoHopFillData>[];
+    }
+
+    public async getTwoHopBuySamplesAsync(
+        sources: ERC20BridgeSource[],
+        takerToken: Address,
+        makerToken: Address,
+        makerAmount: BigNumber,
+    ): Promise<DexSample<TwoHopFillData>[]> {
+        const tokenPaths = this._getTwoHopTokenPaths(takerToken, makerToken);
+        const hopResults = (await Promise.all(tokenPaths.map(async tokenPath =>
+            this._twoHopSampler.getTwoHopBuySampleAsync(sources, tokenPath, makerAmount))));
+        return hopResults.filter(h => !!h) as DexSample<TwoHopFillData>[];
+    }
+
+    private _findSampler(source: ERC20BridgeSource): SourceSampler | undefined {
+        return this._samplers[source];
+    }
+
+    private async _sampleSellsFromSourceAsync(
+        source: ERC20BridgeSource,
+        tokenPath: Address[],
+        takerAmounts: BigNumber[],
+    ): Promise<DexSample[][]> {
+        const sampler = this._findSampler(source);
+        if (!sampler) {
+            return [];
+        }
+        return (await sampler.getSellSamplesAsync(tokenPath, takerAmounts)).filter(s => !!s.length);
+    }
+
+    private async _sampleBuysFromSourceAsync(
+        source: ERC20BridgeSource,
+        tokenPath: Address[],
+        takerAmounts: BigNumber[],
+    ): Promise<DexSample[][]> {
+        const sampler = this._findSampler(source);
+        if (!sampler) {
+            return [];
+        }
+        return (await sampler.getBuySamplesAsync(tokenPath, takerAmounts)).filter(s => !!s.length);
+    }
+
+    private _getExpandedTokenPaths(takerToken: Address, makerToken: Address): Address[][] {
+        return [
+            // A -> B
+            [takerToken, makerToken],
+            // A -> C -> B
+            ...this._getTwoHopTokenPaths(takerToken, makerToken),
+        ].map(p => p.map(a => a.toLowerCase()));
+    }
+
+    private _getTwoHopTokenPaths(takerToken: Address, makerToken: Address): Address[][] {
+        return this._getIntermediateTokens(takerToken, makerToken).map(t => [takerToken, t, makerToken]);
+    }
+
+    private _getIntermediateTokens(takerToken: Address, makerToken: Address): Address[] {
+        return getIntermediateTokens(takerToken, makerToken, this._tokenAdjacencyGraph);
+    }
+}
+
+interface CreateSourceSamplerOpts {
+    chain: Chain;
+    liquidityProviderRegistry?: LiquidityProviderRegistry;
+}
+
+async function createSourceSamplerAsync(source: ERC20BridgeSource, opts: CreateSourceSamplerOpts): Promise<SourceSampler> {
```

Comment:
When you add a new source, don't forget to add to this create function.

Reply:
We can probably use TS to enforce that all `ERC20BridgeSource` values are covered here which would be very neat 🤔

Reply:
Oh good idea. Maybe if I refactor so that this function returns a map of samplers.